### DISCUSSION TITLE
Hypervolume always zero?

### DISCUSSION THREAD

**Discussion by dinohsu1019** [anyoptimization/pymoo#518] [Q&A] [answered]
I am doing NSGA-II with 3 objectives and 12 solution variables.

I am trying to show some performance indicators for each generation, and hypervolume is one of them, but it always shows zero, the following is the code about it, is there anything wrong here? 

Also, my understanding about the indicators: hypervolume: the larger the better; epsilon: the smaller the better, not sure if these are correct?

Thanks in advance.

```
from pymoo.indicators.hv import HV

# 2023-12-06 Show hypervolume for generation NDS, 
# the larger the better among generations
calc_hypervol = HV(ref_point=REF_POINT)

while algorithm.has_next():
    
    algorithm.next()

    n_gen = algorithm.n_gen-1
    n_eval = algorithm.evaluator.n_eval
    opt_F = algorithm.pop.get("F")
    opt_X = algorithm.pop.get("X")
    opt_F, opt_X, opt_index = get_ASF_opt(opt_F, opt_X, WEIGHTS)

    hypervol = calc_hypervol(algorithm.opt.get("F"))
    
    with out_gen_perf:
        print(f"{n_gen=} {n_eval=} {opt_F=} {opt_X} {opt_index=} {hypervol=}")
```

**Comment by blankjul** [comment, marked as answer]
In this example is the `ref_point` dominated by the population? If the `ref_point` dominates the current optimum in the algorithm it will be zero.

**Comment by dinohsu1019** [comment]
> In this example is the `ref_point` dominated by the population? If the `ref_point` dominates the current optimum in the algorithm it will be zero.
I read the description again, it seems I misunderstood the choice of a ref_point, which should be dominated by the solutions instead of dominating. I will try again, thanks.

**Reply by dinohsu1019** [reply]
Just an update. 

In the result I embedded in this post (scroll to the right): https://github.com/anyoptimization/pymoo/discussions/529
I still see all zeros in the hypervolume value, so I haven't yet figured out how to make it work.

**Comment by blankjul** [comment]
Let me give you this example to illustrate when HV is equal to 0 and when it is greater than 0.

```
import numpy as np

from pymoo.algorithms.moo.nsga2 import NSGA2
from pymoo.indicators.hv import HV
from pymoo.problems import get_problem
from pymoo.visualization.scatter import Scatter

problem = get_problem("zdt1")

algorithm = NSGA2(pop_size=200)

solver = algorithm.setup(problem, termination=('n_gen', 50))

# here I define the reference point as (1,1).
# In the beginning it will dominate the entire population: HV = 0
# later when converged some solutions will dominate it: HV > 0
ref_point = np.ones(problem.n_obj)
hv = HV(ref_point=ref_point)

while solver.has_next():
    solver.next()

    # get the objective space values from the optimum
    F = solver.opt.get("F")

    # calculte and print HV
    print(algorithm.n_gen, hv.do(F))

solver.finalize()

res = solver.result()

plot = Scatter()
plot.add(problem.pareto_front(), plot_type="line", color="black", alpha=0.7)
plot.add(res.F, facecolor="none", edgecolor="red")
plot.add(ref_point, color='blue', s=100, marker='x')
plot.show()

```

This shows that HV is zero when it dominates the entire population. As soon as  a solution dominates the references point it will becomes greater than zero.

```
2 0.0
3 0.0
4 0.0
5 0.0
6 0.0
7 0.0
8 0.0
9 0.0
10 0.0
11 0.0
12 0.0
13 0.01325867888623002
14 0.018573696878849057
15 0.02772846861332354
16 0.06470720498913908
17 0.07224132791171181
18 0.10725426346113767
19 0.11313608681091722
20 0.15051301527965627
```

Also look at the plot generated in the end illustration this as well.

**Reply by dinohsu1019** [reply]
Thanks, I will try again.

**Reply by dinohsu1019** [reply]
Following your code with a surely-dominated reference point, I have seen numbers ranging from 0.01 to 900, thanks.

**Comment by dinohsu1019** [comment]
> Let me give you this example to illustrate when HV is equal to 0 and when it is greater than 0.
> 

Quick question: is the variable solver a replacement of the variable algorithm everywhere from the loop? It seems you have only one occurence of algorithm with all the rest solver. I have been using algorithm, so want to confirm, thanks again.

**Reply by blankjul** [reply]
just ignore that I am calling it solver here. It is the same as your code basically.
